CIP-0015/CIP-0015.md

```diff
@@ -33,57 +33,112 @@ A voting key is simply an ED25519 key. How this key is created is up to the wall
-  3: "0x00588e8e1d18cba576a4d35758069fe94e53f638b6faf7c07b8abd2bc5c5cdee47b60edc7772855324c85033c638364214cbfc6627889f81c4"
+  // reward_address - CBOR byte array
+  3: "0x00588e8e1d18cba576a4d35758069fe94e53f638b6faf7c07b8abd2bc5c5cdee47b60edc7772855324c85033c638364214cbfc6627889f81c4",
+  // slot_number
+  4: 5479467
 }
 ```
 
-Signing the voting public key with the staking private key
+The entries under keys 1, 2, 3, and 4 represent the Catalyst voting key,
+the staking key on the Cardano network, the address to receive rewards,
+and the current slot number, respectively. A registration with these metadata
+will be considered valid if the following conditions hold:
+
+- The slot number of the block including this transaction is equal to
+  the slot number value under key 4 or its increment by one.
+- The reward address is a Shelley address discriminated for the same network
+  this transaction is submitted to.
+
+To produce the signature field, the CBOR representation of a map containing
+a single entry with key 61284 and the registration metadata map in the
+format above is formed, designated here as `sign_data`.
+This data is signed with the staking key using one of the two methods,
+discriminated by the key number in the signature metadata map:
+
+1: `sign_data` is signed as is using Ed25519.
+
+2: The blake2b-256 hash of `sign_data` is signed using Ed25519.
+
+Signature example:
 ```
 61285: {
   // signature - ED25119 signature CBOR byte array
   1: "0x8b508822ac89bacb1f9c3a3ef0dc62fd72a0bd3849e2381b17272b68a8f52ea8240dcc855f2264db29a8512bfcd522ab69b982cb011e5f43d0154e72f505f007"
 }
 ```
 
-This corresponds to the following CDDL definition
+### Revocation of a voting key
+
+A previously registered voting key can be revoked by posting another transaction
+with the two metadata fields as above, with the difference that the voting key
+and the reward address fields are absent.
+
+Revocation example:
+```
+61284: {
+  // No voting_key
+  // stake_pub - CBOR byte array
+  2: "0xad4b948699193634a39dd56f779a2951a24779ad52aa7916f6912b8ec4702cee",
+  // slot_number
+  4: 5479468
+}
+```
+
+The signature payload under key 61285 is formed as described above.
+
+### Metadata schema
+
+The format above can be described by the following CDDL definition:
 
 ```
 registration_cbor = {
-  61284: key_registration
-, 61285: registration_signature
+  61284: key_registration / key_revocation,
+  61285: registration_signature
 }
 
 $voting_pub_key /= bytes .size 32
 $staking_pub_key /= bytes .size 32
 $ed25519_signature /= bytes .size 64
-$address /= bytes
+$reward_address /= bytes
+$slot_number /= uint
 
 key_registration = {
-  1 : $voting_pub_key
-, 2 : $staking_pub_key
-, 3 : $address
+  1 : $voting_pub_key,
+  2 : $staking_pub_key,
+  3 : $reward_address,
+  4 : $slot_number,
+}
+
+key_revocation = {
+  2 : $staking_pub_key,
+  4 : $slot_number,
 }
 
 registration_signature = {
-  1 : $ed25519_signature
+  1 : $ed25519_signature //
```

Comment:
here should probable be a comment that this field is not used currently as the signature scheme changed

Reply:
The idea is that either of the schemes can be used interchangeably.

Reply:
Then I think the key should be prefixed with `?` to mark its optionality (https://github.com/input-output-hk/cardano-ledger-specs/blob/2966737bbaf4952a2cf8cd185f253ea5b82d863f/shelley/chain-and-ledger/shelley-spec-ledger-test/cddl-files/shelley.cddl#L56) and explain with a comment that either key 1 or 2 should be specified, as I guess CDDL doesn't have that level of expressivity natively

Reply:
also, do we want to allow both keys at the same time? I can see that leading to hard-to-predict behavior/implementation missalignment for the case when one signature would be valid and the other not